### ISSUE TITLE
A combinator for chunksOf with timeout

### ISSUE THREAD

**Issue by harendra-kumar** [composewell/streamly#456]
We need a combinator that can chunk a stream to fixed size chunks but if the chunk is not completed within a certain time period it will still emit the chunk.

This combinator can useful e.g. to write a stream to stdout or a file but flush it within a certain time period.

I wrote this as stop gap solution, but this is not ideal. It also has an undesired behavior.

```
{-# LANGUAGE FlexibleContexts #-}
module Main (main) where

import Data.Char (ord)
import Data.Maybe (isNothing)
import System.IO (stdin, stdout)

import Streamly
import qualified Streamly.Internal.Prelude as S
import qualified Streamly.Internal.Data.Fold as FL
import qualified Streamly.Internal.Data.Parse as PR
import qualified Streamly.Internal.FileSystem.Handle as FH
import qualified Streamly.Internal.Data.Unicode.Stream as U


{-# INLINE chunksOfTimeout #-}
chunksOfTimeout :: (IsStream t, MonadAsync m)
    => Int -> Double -> FL.Fold m a b -> t m a -> t m b
chunksOfTimeout n t f =
      S.splitWithSuffix isNothing (FL.lcatMaybes f)
    . S.interjectSuffix t (return Nothing)
    . S.intersperseSuffixBySpan n (return Nothing)
    . S.map Just

main = do
      FH.putLines U.encodeLatin1
    $ chunksOfTimeout 10 5 FL.toList
    $ U.decodeLatin1
    $ S.unfold FH.read stdin
```

We also have this unimplemented comment in the FileSystem.Handle module;

```
-- | Write collecting the input in sessions of n seconds or if chunkSize
-- gets exceeded.
{-# INLINE writeByChunksOrSessionsOf #-}
writeByChunksOrSessionsOf :: MonadIO m
    => Int -> Double -> Handle -> SerialT m Word8 -> m ()
writeByChunksOrSessionsOf chunkSize sessionSize h m = undefined
```

This can potentially be implemented as a parser as well.

**Comment by harendra-kumar** [comment]
We also need to think about changing the behavior of default `write` folds to include a timeout. Quite often programmers run into an issue where the data is stuck in the buffer of the `write` routine. If we add a reasonable timeout by default and provide other APIs for better control, it may help avoid such issues.

**Comment by harendra-kumar** [comment]
We can have frame aware buffering combinators such that we can flush the buffer whenever the frame ends. We can have the stream consist of `Just a` values and use a `Nothing` to indicate the end of frame. An end of frame indicates a force flush of the buffer. This will give flushing control to the programmer. Even timeout based combinators can be made from such framed combinators. We just need to inject `Nothing` in the stream at the timeout.

How much performance overhead will such framing introduce, especially for streams of bytes? It may not be much if it fuses well. Should we have such combinators as default to avoid the pitfalls in the current case?

Using the current APIs may cause serious inconvenience. So marking this as high priority. However, breaking changes (changing the behavior of the current API) can only go in the next major release. We can have an alternative API before that though.

**Comment by adithyaov** [comment]
Most of these combinators are going to become much simpler with terminating folds.

**Comment by harendra-kumar** [comment]
fixed #1320
